runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java

```diff
@@ -0,0 +1,412 @@
+   * necessary.
+   *
+   * <p>TODO: This uses deprecated OldDoFn since DoFn does not provide access to state/timer
+   * internals. This should be rewritten to use the <a href="https://s.apache.org/beam-state">State
+   * and Timers API</a> once it is available.
+   */
+  private static class ProcessFn<
+          InputT, RestrictionT, OutputT, TrackerT extends RestrictionTracker<RestrictionT>>
+      extends OldDoFn<KeyedWorkItem<String, ElementRestriction<InputT, RestrictionT>>, OutputT>
+      implements OldDoFn.RequiresWindowAccess {
+    /**
+     * The state cell containing a watermark hold for the output of this {@link DoFn}. The hold is
+     * acquired during the first {@link DoFn.ProcessElement} call for this element and restriction,
+     * and is released when the {@link DoFn.ProcessElement} call returns {@link
+     * DoFn.ProcessContinuation#stop}.
+     *
+     * <p>A hold is needed to avoid letting the output watermark immediately progress together with
+     * the input watermark when the first {@link DoFn.ProcessElement} call for this element
+     * completes.
+     *
+     * <p>The hold is updated with the future output watermark reported by ProcessContinuation.
+     */
+    private static final StateTag<Object, WatermarkHoldState<GlobalWindow>> watermarkHoldTag =
+        StateTags.makeSystemTagInternal(
+            StateTags.<GlobalWindow>watermarkStateInternal(
+                "hold", OutputTimeFns.outputAtLatestInputTimestamp()));
+
+    /**
+     * The state cell containing a copy of the element. Written during the first {@link
+     * DoFn.ProcessElement} call and read during subsequent calls in response to timer firings, when
+     * the original element is no longer available.
+     */
+    private final StateTag<Object, ValueState<WindowedValue<InputT>>> elementTag;
+
+    /**
+     * The state cell containing a restriction representing the unprocessed part of work for this
+     * element.
+     */
+    private StateTag<Object, ValueState<RestrictionT>> restrictionTag;
+
+    private final DoFn<InputT, OutputT> fn;
+    private final Coder<? extends BoundedWindow> windowCoder;
+
+    private transient DoFnInvoker<InputT, OutputT> invoker;
+
+    ProcessFn(
+        DoFn<InputT, OutputT> fn,
+        Coder<InputT> elementCoder,
+        Coder<? extends BoundedWindow> windowCoder) {
+      this.fn = fn;
+      this.windowCoder = windowCoder;
+      elementTag =
+          StateTags.value("element", WindowedValue.getFullCoder(elementCoder, this.windowCoder));
+      DoFnInvoker<InputT, OutputT> invoker = DoFnInvokers.INSTANCE.newByteBuddyInvoker(fn);
+      restrictionTag =
+          StateTags.value("restriction", invoker.<RestrictionT>invokeGetRestrictionCoder());
+    }
+
+    @Override
+    public void setup() throws Exception {
+      invoker = DoFnInvokers.INSTANCE.newByteBuddyInvoker(fn);
+    }
+
+    @Override
+    public void processElement(final ProcessContext c) {
```

Comment:
I spent a few hours trying to do this and couldn't come up with anything remotely clean. This will change a lot in subsequent PRs where I'll add runner hooks into the heart of this method. I'd rather keep it like this for now; I think a linearly structured method of this size is not so bad, and it was definitely more readable than anything else I was able to come up with as a decomposition.